power: move reboot functionality to os lib

gmarull opened the pull request:
Reboot functionality has nothing to do with PM, so move it out to the subsys/os folder. This PR is part of a series of PM codebase cleanup.

gmarull committed:
power: move reboot functionality to os lib

Reboot functionality has nothing to do with PM, so move it out to the
subsys/os folder.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: releases: add notes on reboot header location change

Inform about the reboot header location changes.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
lib: os: reboot: misc cleanup

Remove redundant includes and out-of-date comment.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

martinjaeger commented on lib/os/reboot.c:
```diff
@@ -4,16 +4,8 @@
 #include <kernel.h>
-#include <drivers/timer/system_timer.h>
 #include <sys/printk.h>
-#include <sys/reboot.h>
```
It may work without inclusion of this header, but shouldn't we still keep it because this file implements `sys/reboot.h`?

gmarull replied:
We can keep it, but in this particular case it is not strictly necessary.

martinjaeger reviewed (commented):
The change makes sense.

However, the first two commits should probably be squashed, as otherwise bisecting might fail because the headers are not found anymore.

gmarull commented:
> However, the first two commits should probably be squashed, as otherwise bisecting might fail because the headers are not found anymore.

done

nvlsianpu reviewed (changes requested):
Can power/reboot.h (which redirects to new header) file be provided for backward compatibility?
External project uses this header, this will help switch it smoothly to new on.

gmarull committed:
power: provide old reboot header for compatibility

Make out-of-tree applications transition easier by providing the same
header as before pointing to the new one. Once Zephyr modules are also
migrated this header should produce a deprecation warning if included.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
> Can power/reboot.h (which redirects to new header) file be provided for backward compatibility?
> External project uses this header, this will help switch it smoothly to new on.

good point, done

gmarull commented:
Note: because we have `<power/reboot.h>` users in modules we test I have replaced the warning with a comment until all of them are ported, we need to the migration in two steps.
